### PR TITLE
Add support for pending receipts

### PR DESCRIPTION
**Pull request by hkalodner**
- Add support for answering calls based on the pending state
- Add support for returning transactions receipts for pending transactions
- Add flag to arb-provider-ethers to swap behavior depending on whether messages are deterministic independent of assertions

### CODE REVIEW

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 35
- Added 39
           

Complexity increasing per file
==============================
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
- packages/arb-validator/rollupvalidator/nodeProcessing_test.go  1
- packages/arb-validator-core/evm/evm.pb.go  8
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  5
- packages/arb-validator/rollupvalidator/rpcserver.go  1
         

Complexity decreasing per file
==============================
+ packages/arb-validator-core/evm/evm_advise.go  -1
+ packages/arb-validator-core/evm/tx.go  -4
+ packages/arb-validator-core/evm/fullLog.go  -1
+ packages/arb-validator/rollupvalidator/txtracker.go  -1
+ packages/arb-validator-core/message/delivered.go  -1
         

Clones added
============
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
- packages/arb-validator-core/evm/evm.pb.go  4
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  4
         

Clones removed
==============
+ packages/arb-validator-core/evm/evm_advise.go  -4
+ packages/arb-validator/structures/node_test.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5656453&bid=18517408)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 35
- Added 39
           

Complexity increasing per file
==============================
- packages/arb-validator/rollupvalidator/rpcserver.go  1
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  5
- packages/arb-validator/rollupvalidator/nodeProcessing_test.go  1
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
- packages/arb-validator-core/evm/evm.pb.go  8
         

Complexity decreasing per file
==============================
+ packages/arb-validator-core/evm/tx.go  -4
+ packages/arb-validator-core/evm/evm_advise.go  -1
+ packages/arb-validator-core/message/delivered.go  -1
+ packages/arb-validator/rollupvalidator/txtracker.go  -1
+ packages/arb-validator-core/evm/fullLog.go  -1
         

Clones added
============
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  4
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
- packages/arb-validator-core/evm/evm.pb.go  4
         

Clones removed
==============
+ packages/arb-validator-core/evm/evm_advise.go  -4
+ packages/arb-validator/structures/node_test.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5656453&bid=18517408)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 35
- Added 39
           

Complexity increasing per file
==============================
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  5
- packages/arb-validator/rollupvalidator/rpcserver.go  1
- packages/arb-validator-core/evm/evm.pb.go  8
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
- packages/arb-validator/rollupvalidator/nodeProcessing_test.go  1
         

Complexity decreasing per file
==============================
+ packages/arb-validator-core/evm/tx.go  -4
+ packages/arb-validator-core/evm/evm_advise.go  -1
+ packages/arb-validator/rollupvalidator/txtracker.go  -1
+ packages/arb-validator-core/message/delivered.go  -1
+ packages/arb-validator-core/evm/fullLog.go  -1
         

Clones added
============
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  4
- packages/arb-validator-core/evm/evm.pb.go  4
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
         

Clones removed
==============
+ packages/arb-validator-core/evm/evm_advise.go  -4
+ packages/arb-validator/structures/node_test.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5656453&bid=18517408)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 40
- Added 49
           

Complexity increasing per file
==============================
- packages/arb-validator-core/evm/evm.pb.go  8
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
- packages/arb-validator/rollupvalidator/rpcserver.go  1
- packages/arb-validator/rollupvalidator/nodeProcessing_test.go  1
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  5
         

Complexity decreasing per file
==============================
+ packages/arb-validator-core/evm/tx.go  -4
+ packages/arb-validator-core/evm/evm_advise.go  -1
+ packages/arb-validator-core/message/delivered.go  -1
+ packages/arb-validator/rollupvalidator/txtracker.go  -1
+ packages/arb-validator-core/evm/fullLog.go  -1
         

Clones added
============
- packages/arb-validator-core/evm/evm.pb.go  4
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
- packages/arb-provider-go/validatorProxy.go  2
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  4
         

Clones removed
==============
+ packages/arb-validator-core/evm/evm_advise.go  -4
+ packages/arb-validator/structures/node_test.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5656453&bid=18517408)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 39
- Added 52
           

Complexity increasing per file
==============================
- packages/arb-validator-core/evm/evm.pb.go  8
- packages/arb-validator/rollupvalidator/nodeProcessing_test.go  1
- packages/arb-validator/rollupvalidator/rpcserver.go  1
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  5
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
         

Complexity decreasing per file
==============================
+ packages/arb-validator-core/evm/evm_advise.go  -1
+ packages/arb-validator-core/evm/tx.go  -4
+ packages/arb-validator-core/message/delivered.go  -1
+ packages/arb-validator/rollupvalidator/txtracker.go  -1
+ packages/arb-validator-core/evm/fullLog.go  -1
         

Clones added
============
- packages/arb-validator-core/evm/evm.pb.go  4
- packages/arb-provider-go/validatorProxy.go  2
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  4
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
         

Clones removed
==============
+ packages/arb-validator/structures/node_test.go  -1
+ packages/arb-validator-core/evm/evm_advise.go  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5656453&bid=18517408)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 40
- Added 48
           

Complexity increasing per file
==============================
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  5
- packages/arb-validator-core/evm/evm.pb.go  8
- packages/arb-validator/rollupvalidator/rpcserver.go  1
- packages/arb-validator/rollupvalidator/nodeProcessing_test.go  1
         

Complexity decreasing per file
==============================
+ packages/arb-validator-core/evm/evm_advise.go  -1
+ packages/arb-validator-core/evm/tx.go  -4
+ packages/arb-validator/rollupvalidator/txtracker.go  -1
+ packages/arb-validator-core/evm/fullLog.go  -1
+ packages/arb-validator-core/message/delivered.go  -1
         

Clones added
============
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  4
- packages/arb-provider-go/validatorProxy.go  2
- packages/arb-validator-core/evm/evm.pb.go  4
         

Clones removed
==============
+ packages/arb-validator-core/evm/evm_advise.go  -4
+ packages/arb-validator/structures/node_test.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5656453&bid=18517408)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 40
- Added 48
           

Complexity increasing per file
==============================
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  5
- packages/arb-validator/rollupvalidator/nodeProcessing_test.go  1
- packages/arb-validator/rollupvalidator/rpcserver.go  1
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
- packages/arb-validator-core/evm/evm.pb.go  8
         

Complexity decreasing per file
==============================
+ packages/arb-validator-core/message/delivered.go  -1
+ packages/arb-validator/rollupvalidator/txtracker.go  -1
+ packages/arb-validator-core/evm/fullLog.go  -1
+ packages/arb-validator-core/evm/tx.go  -4
+ packages/arb-validator-core/evm/evm_advise.go  -1
         

Clones added
============
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  4
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
- packages/arb-provider-go/validatorProxy.go  2
- packages/arb-validator-core/evm/evm.pb.go  4
         

Clones removed
==============
+ packages/arb-validator/structures/node_test.go  -1
+ packages/arb-validator-core/evm/evm_advise.go  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5656453&bid=18517408)

File: packages/arb-validator/rollupvalidator/server.go [hunk 1]
```diff
@@ -205,7 +205,22 @@ func (m *Server) CallMessage(ctx context.Context, args *validatorserver.CallMess
...
+	latestBlock := m.man.CurrentBlockId()
+	latestTime := big.NewInt(time.Now().Unix())
+	timeBounds := &protocol.TimeBounds{
+		LowerBoundBlock:     latestBlock.Height,
+		UpperBoundBlock:     latestBlock.Height,
+		LowerBoundTimestamp: latestTime,
+		UpperBoundTimestamp: latestTime,
+	}
+	assertion, steps := mach.ExecuteAssertion(
+		// Call execution is only limited by wall time, so use a massive max steps as an approximation to infinity
```
Comment (edfelten) [inline packages/arb-validator/rollupvalidator/server.go:218]: Do we have to work about DoS attacks here, or should we be charging clients for this service somehow?

Comment (hkalodner) [reply packages/arb-validator/rollupvalidator/server.go:218]: As discussed synchronously, the issue raised here definitely exists, but it also exists in master as well and so should be handled separately from this PR

File: packages/arb-validator/rollupvalidator/rpcserver.go [hunk 2]
```diff
@@ -42,75 +42,101 @@ func NewRPCServer(
...
-	if ret != nil {
-		*reply = *ret
+	if err != nil || ret == nil {
+		return err
+	}
+	reply.RawVal = ret.RawVal
+	return nil
+}
+
+// CallMessage takes a request from a client to process in a temporary context
```
[the remark this answers is not in the document]
Comment (hkalodner) [reply packages/arb-validator/rollupvalidator/rpcserver.go]: Fixed

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 40
- Added 47
           

Complexity increasing per file
==============================
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  5
- packages/arb-validator/rollupvalidator/nodeProcessing_test.go  1
- packages/arb-validator/rollupvalidator/rpcserver.go  1
- packages/arb-validator-core/evm/evm.pb.go  8
         

Complexity decreasing per file
==============================
+ packages/arb-validator-core/evm/fullLog.go  -1
+ packages/arb-validator-core/evm/evm_advise.go  -1
+ packages/arb-validator-core/message/delivered.go  -1
+ packages/arb-validator/rollupvalidator/txtracker.go  -1
+ packages/arb-validator-core/evm/tx.go  -4
         

Clones added
============
- packages/arb-provider-go/validatorProxy.go  2
- packages/arb-validator/rollupvalidator/txtracker_test.go  3
- packages/arb-validator/rollupvalidator/txtrackerdb_test.go  4
- packages/arb-validator-core/evm/evm.pb.go  4
         

Clones removed
==============
+ packages/arb-validator-core/evm/evm_advise.go  -4
+ packages/arb-validator/structures/node_test.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5656453&bid=18517408)

**Pull request merged**